packages/sdk/android/src/main/java/com/primerioreactnative/components/manager/ach/PrimerRNHeadlessUniversalCheckoutStripeAchUserDetailsComponent.kt

```diff
@@ -0,0 +1,310 @@
+          JSONObject().apply {
+            put(
+                "errors",
+                JSONArray().apply { put(JSONObject(json.encodeToString(error.toPrimerErrorRN()))) }
+            )
+          }
+      )
+    }
+  }
+
+  private suspend fun configureStepListener() {
+    component?.componentStep?.collectLatest { step ->
+      when (step) {
+        is AchUserDetailsStep.UserDetailsRetrieved -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_STEP.eventName,
+              JSONObject(json.encodeToString(step.toUserDetailsRetrievedRN()))
+          )
+        }
+        is AchUserDetailsStep.UserDetailsCollected -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_STEP.eventName,
+              JSONObject(json.encodeToString(step.toUserDetailsCollectedRN()))
+          )
+        }
+      }
+    }
+  }
+
+  private suspend fun configureValidationListener() {
+    component?.componentValidationStatus?.collectLatest { validationStatus ->
+      when (validationStatus) {
+        is PrimerValidationStatus.Validating -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_VALIDATING.eventName,
+              JSONObject().apply {
+                putData(validationStatus.collectableData as AchUserDetailsCollectableData)
+              }
+          )
+        }
+        is PrimerValidationStatus.Invalid -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_IN_VALID.eventName,
+              JSONObject().apply {
+                putData(validationStatus.collectableData as AchUserDetailsCollectableData)
+                put(
+                    "errors",
+                    JSONArray(
+                        validationStatus.validationErrors.map {
+                          JSONObject(
+                              json.encodeToString(
+                                  PrimerValidationErrorRN(
+                                      it.errorId,
+                                      it.description,
+                                      it.diagnosticsId,
+                                  )
+                              )
+                          )
+                        }
+                    )
+                )
+              }
+          )
+        }
+        is PrimerValidationStatus.Valid -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_VALID.eventName,
+              JSONObject().apply {
+                putData(validationStatus.collectableData as AchUserDetailsCollectableData)
+              }
+          )
+        }
+        is PrimerValidationStatus.Error -> {
+          sendEvent(
+              PrimerHeadlessUniversalCheckoutComponentEvent.ON_VALIDATION_ERROR.eventName,
+              JSONObject().apply {
+                putData(validationStatus.collectableData as AchUserDetailsCollectableData)
+                put(
```

Comment:
[nit]: we seem to have this in every component, maybe worth centralising this and use in all components?

Reply:
Will do.